### ISSUE TITLE
Rest only High Level APIs completeness

### ISSUE THREAD

**Issue by hub-cap** [elastic/elasticsearch#47678]
There are a number of API that are exposed via REST but not via the Transport Client. They don't necessarily have to be implemented if the goal is feature parity with the Transport Client, yet we should probably have a look at why they were not added to the Transport Client and whether it makes sense to add their support to the high-level REST Client or not. I don't think it makes sense to add support for cat API and ingest processor grok, hence I took them out already.

- [] cluster remote info
- [] get source
- [] delete alias
- [] get upgrade

taken from #27205

**Comment by qianyuxiang01** [comment]
Do you plan to complete those legacy RHLC api? @hub-cap

**Comment by har-vy** [comment]
Hi @hub-cap, I am a beginner, can I work on this?

**Comment by deivids682** [comment]
Hi, @har-vy  do you take a look? I can manage to do this one.

**Comment by har-vy** [comment]
@deivids sure, I would probably have to pass this on.

**Comment by mohamadLy** [comment]
Hi, anyone on this right now. I was thinking picking it up if possible

**Comment by hub-cap** [comment]
I have done the delete alias. You can feel free to choose one of the API's and tell me so I can put your name next to it.

**Comment by mohamadLy** [comment]
Hub-cap thanks for the quick reply. I think I'll go with the source API

**Comment by mohamadLy** [comment]
Hi hub-cap,

I am having issue build the source code. I am trying to build it on ubuntu 18.04. Doesn't seem like java12 exist for ubuntu 18.04. I tried to build with java13 but that did not work either. Do you know any orkaround for this?

I apologize if this is not the right place to post the question.

**Comment by andrewjohnson2** [comment]
Hi, is anyone doing get upgrade? If not I can take it!

**Comment by hub-cap** [comment]
@mohamadLy see https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-to-the-elasticsearch-codebase

@andrewjohnson2 I can mark you down if you like.

**Comment by andrewjohnson2** [comment]
Thanks! I am new to the project and trying to understand some things. To clarify, there is a REST API that exposes get upgrade and I need to implement get upgrade in the high level REST client? Where in the codebase is it currently implemented? Thanks

**Comment by mohamadLy** [comment]
@andrewjohnson2 I am pretty new to this project also, trying to implement get source. Been doing some investigation on how to go about it. This is the path to the source code where you need to make your changes
https://github.com/elastic/elasticsearch/tree/master/client/rest-high-level/src/main/java/org/elasticsearch/client 
I let the others reply to your other question since they are more knowledge about the project then I do. 
Do you know it is possible to import our custom Rest High Level Client Jar on the different project? What I mean is that I build the source and would like to test my implementation. It would be great if can import that Jar to my little test program to test.

**Comment by hrealinho** [comment]
hey guys. first time contributor here, what API would who like me to try and tackle? @hub-cap 
Cheers!

**Comment by andrewjohnson2** [comment]
I did some more research and found this: 
https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestUpgradeStatusAction.java
file which appears to be the existing get upgrade implementation. Is this correct? If so I tried reading the API docs and found that the /_upgrade API has been deprecated? https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-upgrade.html
Any assistance is really appreciated. Thanks!

**Comment by hub-cap** [comment]
So, these are tricky APIs since they are REST only. This means they may not have a specific transport action associated w/ them. The best way to find it is what @andrewjohnson2 has pointed out, which is to find the URI of the action in question,  `_upgrade`, in this case, [which maps to this](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestUpgradeStatusAction.java#L38).

You are right @andrewjohnson2 that this will be removed in the future, but it still exists today, so its still a valid use case for the client.

**Comment by hub-cap** [comment]
@hrealinho you can choose one of the APIs that is not taken above

**Comment by andrewjohnson2** [comment]
@hub-cap thanks so much! For get upgrade specifically, will the implementation be similar in nature to the delete alias implementation? Such as editing similar files and such?

**Comment by hub-cap** [comment]
yes @andrewjohnson2 you should be able to use that as a guide.

**Comment by j-bean** [comment]
Hi, @hub-cap can I start working on `cluster remote info`?

**Comment by jesinity** [comment]
Hi @hub-cap , I can pick "get grok patterns" eventually.

**Comment by andrewjohnson2** [comment]
I won't be able to complete get upgrade. Thanks for your assistance

**Comment by timoninmaxim** [comment]
Hi @hub-cap , could I work on "get upgrade" issue instead?

> I won't be able to complete get upgrade. Thanks for your assistance

**Comment by martijnvg** [comment]
@thomdixon Yes, you can work on that issue.

**Comment by martijnvg** [comment]
@jesinity I've assigned get grok patterns to you. Thanks!

**Comment by timoninmaxim** [comment]
hi @martijnvg @hub-cap 

I made some research for Upgrade API. Under the hood it just recreate segments with upper version of lucene index. But sometimes that is not enough to upgrade index correctly (for example, [#17446](https://github.com/elastic/elasticsearch/issues/17446)). And _I think_ this is a reason why the API considered to be removed in favor to reindex API. Unfortunately, I failed to find any topic with such discussion, or topic that describes why TransportClient didn't implement the api.

Also there was an issue with tests are skipped for the API as it marked for removal: [#25311](https://github.com/elastic/elasticsearch/issues/25311)

I wonder why is not the API marked as deprecated at least? Is it a good moment to completely remove it instead of bringing a new code?

**Comment by martijnvg** [comment]
@timoninmaxim Good point. I don't recall where this discussion took place and almost forgot about it. The reference documentation page about upgrade api describes that it is [deprecated](https://www.elastic.co/guide/en/elasticsearch/reference/7.5/indices-upgrade.html).

So let's not add HLRC support for this API. I will scratch it from this list.

**Comment by timoninmaxim** [comment]
@martijnvg good! Buy marking as deprecated I mean that in the [rest-api-spec](https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/api/indices.upgrade.json) the endpoint isn't marked as deprecated.

**Comment by martijnvg** [comment]
@timoninmaxim Agreed, this api should be marked as deprecated in the specs and also return a deprecation warning header if used.

**Comment by timoninmaxim** [comment]
@martijnvg ok, I'll have a look

**Comment by timoninmaxim** [comment]
Hi @hub-cap  @martijnvg  !
I've prepared a PR (see above) but there is no response for 16 days. Is there something I should add to it?

**Comment by timoninmaxim** [comment]
Hi @hub-cap @martijnvg !
I emailed @mohamadLy about if it possible to take "get source" action instead him as it is already weeks without any updates. He сonfirmed that. So, if you dont have objections I would complete this API.

**Comment by martijnvg** [comment]
Hey @timoninmaxim, Thanks for volunteering! I will updated the issue description. As you may already have noticed, Lee is reviewing your deprecate _upgrade API.

**Comment by timoninmaxim** [comment]
Hi @martijnvg ! Please check my PR #50885 for "get source" API . Sorry, I didn't find how to request review in the github UI.

**Comment by timoninmaxim** [comment]
@martijnvg thanks a lot for the help with the tasks. It were my first PRs ever to opensource projects!

**Comment by martijnvg** [comment]
@timoninmaxim You're welcome and thank you for picking up these tasks!

**Comment by Priya-Raut** [comment]
Is get grok patterns API still available to start working on it?

**Comment by jesinity** [comment]
It is available if you want ti pick it

**Comment by Priya-Raut** [comment]
Thanks! I will start working on it then.

**Comment by Priya-Raut** [comment]
Hi,
I am trying to contribute to elasticsearch for the first time. I am not able to run the code yet.
After following instructions from [CONTRIBUTING.md](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-to-the-elasticsearch-codebase), 
1. JAVA_HOME is set to JDK 14
2. The project is imported in Intellij 2020.1
3. Docker is not set up yet
4. Ran the command ./gradlew :run

However, it is running forever. 
Can someone please point out if I am doing something wrong?

**Comment by tvernum** [comment]
> However, it is running forever.

What does "forever" mean exactly?
Elasticsearch is big, and it takes a fair amount of time to compile, but it should not take _hours_ on modern hardware.

Is the build running, but slowly? Or has it stalled completely (and if so, where)?

**Comment by Priya-Raut** [comment]
> What does "forever" mean exactly?
Right! So, It has been running for about 2 hours now.

> Is the build running, but slowly? Or has it stalled completely (and if so, where)?
It has stalled at this line
[2020-05-25T09:08:20,481][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [ml-size-based-ilm-policy]

Should I put the log file here?

**Comment by tvernum** [comment]
Aha! It's not actually stalled, it's running and waiting for you to do stuff.

I'll open an issue to clarify the CONTRIBUTING instructions - it's definitely not clear from that doc or from the build output that it's at the point where you can actually connect.

At this point you can just connect via curl (in a different terminal session) with
```
curl -u elastic:password localhost:9200
```

**Comment by Priya-Raut** [comment]
@tvernum Thanks a lot! It means I was waiting for nothing ><
I got the response! yay!
{
  "name" : "runTask-0",
  "cluster_name" : "runTask",
  "cluster_uuid" : "8FnkV_YLSxmkhxB0yyVYBw",
  "version" : {
    "number" : "8.0.0-SNAPSHOT",
    "build_flavor" : "default",
    "build_type" : "tar",
    "build_hash" : "2a8b5787466c1e2e478c266db76bd2fba580c1ce",
    "build_date" : "2020-05-25T05:37:19.699441Z",
    "build_snapshot" : true,
    "lucene_version" : "8.6.0",
    "minimum_wire_compatibility_version" : "7.9.0",
    "minimum_index_compatibility_version" : "7.0.0"
  },
  "tagline" : "You Know, for Search"
}

**Comment by Priya-Raut** [comment]
I have started working on the get grok patterns (GET /_ingest/processor/grok) API call to include it in the client.
Here are a few queries from me: 

1. Just to confirm is this the API we need include in the client?
https://www.elastic.co/guide/en/elasticsearch/reference/current/grok-processor.html#grok-processor-rest-get

2. I am planning to put the code in [RestHighLevelClient.java](https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java)
Is it ok? If not, where should I move it?

3. Where should the client-side documentation be placed under this location?
https://www.elastic.co/guide/en/elasticsearch/client/java-rest/7.7/java-rest-high.html

**Comment by sethmlarson** [comment]
Hey! So back in https://github.com/elastic/elasticsearch/issues/27205 there were APIs that were left off the list (because they expose `ShardRouting` and `ClusterState`) and that information didn't make the leap to this issue. Is there any plan on implementing these APIs? cc @e-mars

**Comment by Chr0nicl3** [comment]
@Priya-Raut are you still working on the solution?
If not, I'd like to work on this issue.

**Comment by jtsa0267** [comment]
Hi @hub-cap and @danhermann what is the state of this issue? I see several people showed interest in working on this problem but those were months ago.

> 
> 
> [...] I don't think it makes sense to add support for [...] ingest processor grok, hence I took them out already.
> 
>     * [x]  cluster remote info (@j-bean)
> 
>     * [x]  get source (@timoninmaxim)
> 
>     * [x]  delete alias (@hub-cap)
>       ~- [ ] get upgrade (@timoninmaxim)~
> 
>     * [ ]  get grok patterns (GET /_ingest/processor/grok) (@jesinity)
> 
> 
> taken from #27205

I'm not sure I'm following, @hub-cap mentioned that he didn't think there should be support (I'm assuming support as in exposing `GET /_ingest/processor/grok` as an API endpoint?) for the grok processor, but at the same time it is listed as an unfinished to-do item at the bottom of the list. 

Secondly, assuming get-grok-patterns is still a feature waiting to be implemented. I can already run `curl -XGET localhost:9200/_ingest/processor/grok?pretty` to retrieve a series of patterns, so what exactly is the issue here?

Please let me know if I'm on the right track, new to this project and just trying to understand the codebase, any pointers would be appreciated.

**Comment by timDINOS** [comment]
Hi @hub-cap . Is this issue still open? I would like to work on this issue.

**Comment by dakrone** [comment]
The Java Rest High Level Client has been deprecated (and removed) in later versions of Elasticsearch, in favor of the official Elasticsearch Java Client — https://github.com/elastic/elasticsearch-java

I'm going to close this issue since it is no longer needed.
